### PR TITLE
Remove cctor from BitConverter

### PR DESCRIPTION
**Pull request by EgorBo**
will it work with reflection?

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 1]
```diff
@@ -669,25 +665,24 @@ public static string ToString(byte[] value, int startIndex, int length)
             // (int.MaxValue / 3) == 715,827,882 Bytes == 699 MB
             ArgumentOutOfRangeException.ThrowIfGreaterThan(length, int.MaxValue / 3);
 
-            return string.Create(length * 3 - 1, (value, startIndex, length), static (dst, state) =>
+            int stringLength = length * 3 - 1;
+            string result = string.FastAllocateString(stringLength);
+
+            var dst = new Span<char>(ref result.GetRawStringData(), stringLength);
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:671]: Just to make it a bit more obvious in the span ctor that the length is correct:
```suggestion
            string result = string.FastAllocateString(length * 3 - 1);

            var dst = new Span<char>(ref result.GetRawStringData(), result.Length);
```

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 2]
```diff
@@ -669,25 +665,24 @@ public static string ToString(byte[] value, int startIndex, int length)
...
-                while (i < src.Length)
-                {
-                    b = src[i++];
-                    dst[j++] = '-';
-                    dst[j++] = HexConverter.ToCharUpper(b >> 4);
-                    dst[j++] = HexConverter.ToCharUpper(b);
-                }
-            });
+            }
+            return result;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:685]: Nit:
```suggestion

            return result;
```

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 3]
```diff
@@ -669,25 +665,24 @@ public static string ToString(byte[] value, int startIndex, int length)
...
+            dst[j++] = HexConverter.ToCharUpper(b);
+            while (i < src.Length)
             {
-                var src = new ReadOnlySpan<byte>(state.value, state.startIndex, state.length);
-
-                int i = 0;
-                int j = 0;
-
-                byte b = src[i++];
+                b = src[i++];
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:680]: This is one of those places where even if we changed it to:
```C#
while ((uint)i < (uint)str.Length)
```
the JIT today still wouldn't remove the bounds check for the subsequent src access, right?

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 4]
```diff
@@ -16,13 +16,9 @@ public static class BitConverter
...
         // The value is set to true if the architecture is
         // little endian; false if it is big endian.
-#if BIGENDIAN
         [Intrinsic]
-        public static readonly bool IsLittleEndian /* = false */;
-#else
-        [Intrinsic]
-        public static readonly bool IsLittleEndian = true;
-#endif
+        // It's an intrinsic in all runtimes and doesn't need a static constructor
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:20]:
```suggestion
        // It's an intrinsic in all runtimes and so doesn't need to be initialized here.
        // Not adding such initialization helps avoid BitConverter getting a static cctor.
```

**Review by stephentoub — APPROVED** [review]

**Comment by EgorBo** [comment]
I feel like it won't work as is with Reflection and there has to be an additional logic to set the actual field in memory to `1` just like it's done with String.Empty

**Comment by stephentoub** [comment]
> I feel like it won't work as is with Reflection

You could add a test :)

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs [hunk 5]
```diff
@@ -17,7 +17,8 @@ public static class BitConverter
         // The value is set to true if the architecture is
         // little endian; false if it is big endian.
         [Intrinsic]
-        // It's an intrinsic in all runtimes and doesn't need a static constructor
+        // It's an intrinsic in all runtimes and so doesn't need to be initialized here.
+        // Not adding such initialization helps avoid BitConverter getting a static cctor.
         public static readonly bool IsLittleEndian;
```
Comment (tannergooding) [inline src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:22]: nit: Can we add a test that validates reflection works as expected?

**Review by jkotas — CHANGES_REQUESTED** [review]
This does not work for native AOT and it adds extra useless work during startup. I do not think that this is a good change.

What is the problem that this is trying to solve?

**Comment by stephentoub** [comment]
> I do not think that this is a good change.

I'd still keep the test and the changes to BitConverter.ToString.

**Comment by EgorBo** [comment]
> This does not work for native AOT and it adds extra useless work during startup. I do not think that this is a good change.
> 
> What is the problem that this is trying to solve?

Yeah, when I found out that it needs VM-side changes to fill that field I realized it's not worth the effort

> I'd still keep the test and the changes to BitConverter.ToString.

Reverted all changes but those

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs (hunk 5 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs]: Added

File: src/libraries/System.Private.CoreLib/src/System/BitConverter.cs (hunk 3 above)
Comment (EgorBo) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:683]: I'll see what I can do with this in net8.0 at some point, but loop-related bound checks have always been the worst (fragile) 😥

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/BitConverter.cs:683]: I opened https://github.com/dotnet/runtime/issues/83349 :)

**Pull request merged**